supernode/daemon/mgr/cdn/cdn_gc.go

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright The Dragonfly Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cdn
+
+import (
+	"context"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/dragonflyoss/Dragonfly/pkg/errortypes"
+	"github.com/dragonflyoss/Dragonfly/supernode/config"
+	"github.com/dragonflyoss/Dragonfly/supernode/daemon/mgr"
+	"github.com/dragonflyoss/Dragonfly/supernode/store"
+
+	"github.com/emirpasic/gods/maps/treemap"
+	godsutils "github.com/emirpasic/gods/utils"
+	"github.com/sirupsen/logrus"
+)
+
+func (cm *Manager) GetGCTaskIDs(ctx context.Context, taskMgr mgr.TaskMgr) ([]string, error) {
+	var gcTaskIDs []string
+
+	freeDisk, err := cm.cacheStore.GetAvailSpace(ctx, getHomeRawFunc())
+	if err != nil {
+		return nil, err
+	}
+	if freeDisk > cm.cfg.YoungGCThreshold {
+		return nil, nil
+	}
+
+	fullGC := false
+	if freeDisk <= cm.cfg.FullGCThreshold {
+		fullGC = true
+	}
+	logrus.Debugf("start to exec gc with fullGC: %t", fullGC)
+
+	gapTasks := treemap.NewWith(godsutils.Int64Comparator)
+	intervalTasks := treemap.NewWith(godsutils.Int64Comparator)
+
+	// walkTaskIDs is used to avoid processing multiple times for the same taskID
+	// which is extracted from file name.
+	walkTaskIDs := make(map[string]bool, 0)
+	walkFn := func(path string, info os.FileInfo, err error) error {
+		logrus.Debugf("start to walk path(%s)", path)
+
+		if err != nil {
+			logrus.Errorf("failed to access path(%s): %v", path, err)
+			return err
+		}
+		if info.IsDir() {
+			logrus.Warnf("file is not regular file(%s)", path)
```

Comment:
delete this log. There're many sub-directories in download home.

Reply:
Done.